modules/k3s/k3s.go

```diff
@@ -0,0 +1,202 @@
+package k3s
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"path/filepath"
+
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/mount"
+	"github.com/docker/go-connections/nat"
+	"github.com/testcontainers/testcontainers-go"
+	"github.com/testcontainers/testcontainers-go/wait"
+	"gopkg.in/yaml.v3"
+	//kubeCon "github.com/giantswarm/kubeconfig/"
+)
+
+var (
+	defaultKubeSecurePort     = "6443/tcp"
+	defaultRancherWebhookPort = "8443/tcp"
+	// kubeConfigYaml            string
+)
+
+// K3sContainer represents the K3s container type used in the module
+type K3sContainer struct {
+	testcontainers.Container
+}
+
+// RunContainer creates an instance of the K3s container type
+func RunContainer(ctx context.Context, opts ...testcontainers.ContainerCustomizer) (*K3sContainer, error) {
+	req := testcontainers.ContainerRequest{
+		Image: "docker.io/rancher/k3s:latest",
+		ExposedPorts: []string{
+			defaultKubeSecurePort,
+			defaultRancherWebhookPort,
+		},
+		Privileged: true,
+		HostConfigModifier: func(hc *container.HostConfig) {
+			hc.CgroupnsMode = "host"
+			hc.Tmpfs = map[string]string{
+				"/run":     "",
+				"/var/run": "",
+			}
+			hc.Mounts = []mount.Mount{}
+
+		},
+		Cmd: []string{
+			"server",
+			"--disable=traefik",
+			"--tls-san=localhost",
+		},
+		Env: map[string]string{
+			"K3S_KUBECONFIG_MODE": "644",
+		},
+		WaitingFor: wait.ForLog("k3s is up and running"),
+	}
+
+	genericContainerReq := testcontainers.GenericContainerRequest{
+		ContainerRequest: req,
+		Started:          true,
+	}
+
+	for _, opt := range opts {
+		opt.Customize(&genericContainerReq)
+	}
+
+	container, err := testcontainers.GenericContainer(ctx, genericContainerReq)
+	if err != nil {
+		return nil, err
+	}
+
+	return &K3sContainer{Container: container}, nil
+}
+
+func WithNetwork(networks []string) testcontainers.CustomizeRequestOption {
+	return func(req *testcontainers.GenericContainerRequest) {
+		req.Networks = networks
+	}
+}
+
+func WithNetworkAlias(networkAlias map[string][]string) testcontainers.CustomizeRequestOption {
+	return func(req *testcontainers.GenericContainerRequest) {
+		req.NetworkAliases = networkAlias
+	}
+}
+
+func WithKubectlConfigFile(configFile string) testcontainers.CustomizeRequestOption {
+	return func(req *testcontainers.GenericContainerRequest) {
+		cf := testcontainers.ContainerFile{
+			HostFilePath:      configFile,
+			ContainerFilePath: filepath.Join(".", "config"),
+			FileMode:          0755,
+		}
+		req.Files = append(req.Files, cf)
+	}
+}
+
+func WithCmd(commands []string) testcontainers.CustomizeRequestOption {
+	return func(req *testcontainers.GenericContainerRequest) {
+		req.Cmd = commands
+	}
+}
+
+// getkubeConfigYaml returns the string of modifed kubeconfig with server url
+func (c *K3sContainer) getkubeConfigYaml(ctx context.Context) (string, error) {
```

Comment:
Sounds good only, I guess, since they will be mostly used in tests, exposing it shouldn't be a problem and yes since mostly it might be consumed as bytes (like clientcmd of k8s) so I will return it as that

Reply:
That's exactly the use case: to consume that method in the tests! As users will use the module in their tests :) Therefore, the modules should expose that functionality that is useful for the users.

With that in mind, I can imagine developers willing to get the kubeconfig from the cluster in order to be able to interact with it using that file.

Reply:
Yes, totally makes sense, just love how easy it is going to make for them! Thanks for this!